### PR TITLE
Introduce file_id on Windows

### PR DESCRIPTION
**Pull request by JunkuiZhang**
Added a function `file_id` to get the file id on windows, which is similar to inode on unix. 

Release Notes:

- N/A

### CODE REVIEW

**Comment by JunkuiZhang** [comment]
@cla-bot check

File: crates/fs/src/fs.rs [hunk 1]
```diff
@@ -1337,6 +1336,34 @@ pub fn copy_recursive<'a>(
...
+    use std::os::windows::{fs::OpenOptionsExt, io::AsRawHandle};
+
+    use windows_sys::Win32::{
+        Foundation::HANDLE,
+        Storage::FileSystem::{
+            GetFileInformationByHandle, BY_HANDLE_FILE_INFORMATION, FILE_FLAG_BACKUP_SEMANTICS,
+        },
+    };
+
+    let file = std::fs::OpenOptions::new()
```
Comment (mikayla-maki) [inline crates/fs/src/fs.rs:1350]: We shouldn't be doing synchronous file opening here, this needs to be swapped to use `smol::fs`

File: crates/fs/src/fs.rs [hunk 2]
```diff
@@ -245,9 +245,8 @@ impl Fs for RealFs {
         #[cfg(unix)]
         let inode = metadata.ino();
 
-        // todo!("windows")
         #[cfg(windows)]
-        let inode = 0;
+        let inode = file_id(path);
```
Comment (mikayla-maki) [inline crates/fs/src/fs.rs:249]: We already open the file above using smol, can we restructure this code somehow to not open the file twice?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Review by mikayla-maki — CHANGES_REQUESTED** [review]
Looks good! Just a few small changes :)

File: crates/fs/src/fs.rs [hunk 3]
```diff
@@ -1337,6 +1336,39 @@ pub fn copy_recursive<'a>(
...
+        .read(true)
+        .custom_flags(FILE_FLAG_BACKUP_SEMANTICS)
+        .open(path)
+        .await
+        .expect("Getting file id error: unable to open file!");
+
+    let mut info: BY_HANDLE_FILE_INFORMATION = unsafe { std::mem::zeroed() };
+    // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileinformationbyhandle
+    // This function supports Windows XP+
+    let ret = unsafe { GetFileInformationByHandle(file.as_raw_handle() as HANDLE, &mut info) };
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/fs/src/fs.rs]: I'm sorry, I'm not particularly familiar with async. Thank you very much for your review; it has taught me a lot!

File: crates/fs/src/fs.rs [hunk 4]
```diff
@@ -1337,6 +1336,39 @@ pub fn copy_recursive<'a>(
     .boxed()
 }
 
+// todo!(windows)
+// can we get file id not open the file twice?
+// https://github.com/rust-lang/rust/issues/63010
+#[cfg(target_os = "windows")]
+async fn file_id(path: impl AsRef<Path>) -> u64 {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/fs/src/fs.rs]: Your review has given me a much better understanding of what truly constitutes good code, and what good codes should be like, thx!

**Review by mikayla-maki — APPROVED** [review]

**Pull request merged**